File: python/minscheme/parser.py
import re
from typing import List, Tuple

from minscheme import data

_SKIP_RE = re.compile(r'^(\s+|;.*)+')
_TOKEN_RE = re.compile(r'^[^\s);]+')
_NUM_RE = re.compile(r'^-?[0-9]+$')


class ParseError(Exception):
    pass


def _make_quote(value: data.Value) -> data.Pair:
    return data.from_native_list([data.Symbol('quote'), value])


def _parse_skip(code: str) -> str:
    return _SKIP_RE.sub('', code)


def _parse_value(code: str) -> Tuple[data.Value, str]:
    if code.startswith('\''):
        value, rest_code = _parse_value(code[1:])
        return _make_quote(value), rest_code

    if code.startswith('('):
        values, rest_code = _parse_list(code[1:])
        rest_code = _parse_skip(rest_code)
        if not rest_code.startswith(')'):
            raise ParseError(rest_code)
        return data.from_native_list(values), rest_code[1:]

    match = _TOKEN_RE.search(code)
    if not match:
        raise ParseError(code)
    token = match.group()
    if _NUM_RE.search(token):
        value = data.Int(int(token))
    elif token == '#f':
        value = data.Bool(False)
    elif token == '#t':
        value = data.Bool(True)
    else:
        value = data.Symbol(token)
    return value, code[match.end():]


def _parse_list(code: str) -> Tuple[List[data.Value], str]:
    values = []
    while True:
        code = _parse_skip(code)
        if not code or code.startswith(')'):
            break
        value, code = _parse_value(code)
        values.append(value)
    return values, code


def parse(code: str) -> List[data.Value]:
    values, excess_code = _parse_list(code)
    if excess_code:
        raise ParseError(excess_code)
    return values

File: haskell/app/MinScheme/Parser.hs
module MinScheme.Parser( parse ) where

import Data.List
import Data.Maybe
import MinScheme.Data
import Text.Regex

parseSkip :: String -> String
parseSkip code =
  let skipped = head $ fromJust $ matchRegex skipRe code in
  drop (length skipped) code
  where
    skipRe = mkRegex "(^([ \t\r\n]|;.*)*)"

makeQuote :: Value -> IO Value
makeQuote value =
  valueFromList [Symbol "quote", value]

parseValue :: String -> IO (Value, String)
parseValue code =
  case code of
    _ | "'" `isPrefixOf` code -> do
      (value, code') <- parseValue (tail code)
      quoted <- makeQuote value
      return (quoted, code')
    _ | "(" `isPrefixOf` code -> do
      (values, code') <- parseList (tail code)
      let code'' = parseSkip code'
      if ")" `isPrefixOf` code''
        then do
          values' <- valueFromList values
          return (values', tail code'')
        else error $ "parse error: malformed list: " ++ code''
    _ ->
      case matchRegex tokenRe code of
        Nothing -> error $ "parse error: malformed token: " ++ code
        Just (token:_) ->
          let
            code' = drop (length token) code
            value = case token of
              _ | isJust (matchRegex numRe token) -> Integer (read token)
              "#t" -> Boolean True
              "#f" -> Boolean False
              _ -> Symbol token
          in
            return (value, code')
        _ -> undefined
  where
    tokenRe = mkRegex "(^[^ \t\r\n);]+)"
    numRe = mkRegex "(-?[0-9]+$)"

parseList :: String -> IO ([Value], String)
parseList code =
  let code' = parseSkip code in
  case code' of
    "" -> return ([], code')
    _ | ")" `isPrefixOf` code' -> return ([], code')
    _ -> do
      (value, code'') <- parseValue code'
      (values, code''') <- parseList code''
      return (value : values, code''')

parse :: String -> IO [Value]
parse code = do
  (exprs, code') <- parseList code
  if null code' then return exprs else error $ "parse error: excessive code: " ++ code'
